os/arch/arm/src/amebasmart/amebasmart_idle.c

```diff
@@ -23,14 +23,160 @@
  ****************************************************************************/
 
 #include <tinyara/config.h>
-
 #include <tinyara/arch.h>
-#include "up_internal.h"
+#include <tinyara/irq.h>
+
+#ifdef CONFIG_PM
+#include <tinyara/pm/pm.h>
+#include "amebasmart_config.h"
+#include "arch_timer.h"
+#include "ameba_soc.h"
+#include "osdep_service.h"
+#endif
+
+#ifdef CONFIG_SMP
+#include "gic.h"
+#endif
 
 /****************************************************************************
- * Public Functions
+ * Name: up_idlepm
+ *
+ * Description:
+ *   Perform IDLE state power management.
+ *
  ****************************************************************************/
 
+#ifdef CONFIG_PM
+static enum pm_state_e oldstate = PM_NORMAL;
+static void up_idlepm(void)
+{
+	uint32_t xModifiableIdleTime = 0;
+	enum pm_state_e newstate;
+	irqstate_t flags;
+	int ret;
+
+	/* Decide, which power saving level can be obtained */
+	newstate = pm_checkstate(PM_IDLE_DOMAIN);
+
+	/* Check for state changes */
+	if (newstate != oldstate)
+	{
+		//TODO: Critical section code needed for SMP case?
+		//Any additional implications of putting a core in critical section while trying to sleep?
+		/* Perform board-specific, state-dependent logic here */
+	  	pmvdbg("newstate= %d oldstate=%d\n", newstate, oldstate);
+
+		/* Then force the global state change */
+		ret = pm_changestate(PM_IDLE_DOMAIN, newstate);
+		if (ret < 0) {
+			/* The new state change failed, revert to the preceding state */
+			pmdbg("State change failed! Current state = %d, newstate = %d\n", oldstate, newstate);
+			newstate = oldstate;
+		} else {
+			/* Save the new state */
+			oldstate = newstate;
+		}
+		/* MCU-specific power management logic */
+		switch (newstate) {
+			case PM_NORMAL:
+				break;
+			case PM_IDLE:
+				break;
+			case PM_STANDBY:
+				break;
+			case PM_SLEEP:
+				ap_timer_helper();
```

Comment:
If the state is able to fall into PM_SLEEP, it is impossible that the sleep will be blocked again, as the remaining operation is to setup the pre-sleep logic flow (Meaning that the driver state rejection/ application usage protection should be robust, as there is no blocker left after reaching PM_SLEEP, THE ONLY blocker is in SMP, which if secondary core is not in hotplug mode, we will set it to hotplug mode first, and go to sleep at the next cycle, but since this feature is not fully completed yet, we should discuss it at later stage).
Thus the timer can be handled properly as for single core case, I will add in some consideration logic to setup timer for condition with or w/o SMP.

Reply:
Ohkay, I get your point.
Could you add a TODO comment here please to indicate that we have to check some things in the future as you mentioned?

Reply:
I have added some TODO comment at below section, I think that part should be relevant to this topic.

Reply:
I saw the comment below for SMP case and timer.
However, I still request you to add a short TODO comment here for tracking purposes please.